In [1]:
from sklearn.datasets import load_diabetes
x,y=load_diabetes(return_X_y=True)

# here I load diabetes dataset from detasets class provided by sci-kit learn

In [2]:
print(x[:5,:])
print(x.shape)

[[ 0.03807591  0.05068012  0.06169621  0.02187239 -0.0442235  -0.03482076
  -0.04340085 -0.00259226  0.01990749 -0.01764613]
 [-0.00188202 -0.04464164 -0.05147406 -0.02632753 -0.00844872 -0.01916334
   0.07441156 -0.03949338 -0.06833155 -0.09220405]
 [ 0.08529891  0.05068012  0.04445121 -0.00567042 -0.04559945 -0.03419447
  -0.03235593 -0.00259226  0.00286131 -0.02593034]
 [-0.08906294 -0.04464164 -0.01159501 -0.03665608  0.01219057  0.02499059
  -0.03603757  0.03430886  0.02268774 -0.00936191]
 [ 0.00538306 -0.04464164 -0.03638469  0.02187239  0.00393485  0.01559614
   0.00814208 -0.00259226 -0.03198764 -0.04664087]]
(442, 10)


In [3]:
print(y[:5])
print(y.shape)

[151.  75. 141. 206. 135.]
(442,)


#  Y^ = X * B
1. y^ is the matrix containing predicted values of dependent variable
> size of y^ is n * 1 where n is the no of rows/persons/indivudual
> in our case it is 442 * 1
2. X is the matrix contain the values of independent variables 
> size of X matris is n * m where n is the no of rows/ persons/ individual and m is the dimention of data
> in our case n=442 and m=11 (total no of columns in our dataset including y)   so it is   442 * 11
3. B is the matrix that was predicted by model containing bius(B0) and weights(B1,B2,B3...,Bm) size is m * 1
> in our case size of B is 11 * 1

* **For calculate B we derived the formula: **
# B = (X * X^T)^-1 (X^T * Y)

***FOR DERIVATION VISIT MY GITHUB REPOSITORY WHRE YOU CAN FIND THE IDEA ABOUT IT IN "Backend Of Linear Regression" FILE.***

In [4]:
# for finding transpose: X^T = X.T
# for finding inverse: numpy.linalg.inv() function is used
# for finding scaler product of metrics we use np.dot() fuction

* as we see in the "Backend Of Linear Regression" there always on extra column instead of inpupt column whose value is 1.
* it is important because constant/bius (B0) is muliply with 1 and help to get the equeation y^1 = B0 + B1x11 + B2X12 + ...
* but in our data set column containing that 1's is not present so we have to add in out AI_Linear_Regression class during model fitting

* Below is the intution of How we add 1's value containing column at index 0.

In [5]:
import numpy as np

In [6]:
print(x)

#before adding extra column

[[ 0.03807591  0.05068012  0.06169621 ... -0.00259226  0.01990749
  -0.01764613]
 [-0.00188202 -0.04464164 -0.05147406 ... -0.03949338 -0.06833155
  -0.09220405]
 [ 0.08529891  0.05068012  0.04445121 ... -0.00259226  0.00286131
  -0.02593034]
 ...
 [ 0.04170844  0.05068012 -0.01590626 ... -0.01107952 -0.04688253
   0.01549073]
 [-0.04547248 -0.04464164  0.03906215 ...  0.02655962  0.04452873
  -0.02593034]
 [-0.04547248 -0.04464164 -0.0730303  ... -0.03949338 -0.00422151
   0.00306441]]


In [7]:
np.insert(x,0,1,axis=1)

#axis = 0 means x axis and axis = 1 means y axis
#note:- np.insert() function not change original data, so don't worry it not modifies x

array([[ 1.        ,  0.03807591,  0.05068012, ..., -0.00259226,
         0.01990749, -0.01764613],
       [ 1.        , -0.00188202, -0.04464164, ..., -0.03949338,
        -0.06833155, -0.09220405],
       [ 1.        ,  0.08529891,  0.05068012, ..., -0.00259226,
         0.00286131, -0.02593034],
       ...,
       [ 1.        ,  0.04170844,  0.05068012, ..., -0.01107952,
        -0.04688253,  0.01549073],
       [ 1.        , -0.04547248, -0.04464164, ...,  0.02655962,
         0.04452873, -0.02593034],
       [ 1.        , -0.04547248, -0.04464164, ..., -0.03949338,
        -0.00422151,  0.00306441]])

In [8]:
class AI_Linear_Regression:
    
    def __init__(self):
        self.coef_ = None
        self.intercept_ = None
        self.B = None
        
    def fit(self,x,y):
        x=np.insert(x,0,1,axis=1)
        self.B = np.linalg.inv(np.dot(x.T,x)).dot(x.T).dot(y)
        self.intercept_ = self.B[0]
        self.coef_ = self.B[1:]
    
    def predict(self,x):
        y_predict=np.dot(x,self.coef_) + self.intercept_
        print(y_predict)
        
    

In [9]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=11)

#split training data and testing data using train_test_split function of odel_selection class provided by sci-kit librabry.

In [10]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(353, 10)
(353,)
(89, 10)
(89,)


In [11]:
model=AI_Linear_Regression()

#making object of our class

In [12]:
model.fit(x_train,y_train)

In [13]:
model.predict(x_test)

#testing our model

[116.0566433   82.3085256  160.86715899 189.18169509  75.32179675
 231.1964373   96.22092022 253.17304788 174.1389304   96.2997225
 156.32104182 262.81516649  88.22905626 122.14265396  67.34337199
  86.31433427  82.71228406 152.04150296 179.50392491  96.64118319
  59.77455683 191.89591316 108.09513664  56.8553631  191.87763873
 154.55060305 213.40366806 182.26534397 255.41664825 155.36595373
 186.88366323 148.76871858 168.15381349 179.7810158  128.00761655
 127.01136478  85.85344325 183.79617317 138.85457876 169.50035675
  85.71119008 104.20880048 200.00938972 216.6808666  193.73349677
 165.87641058 183.586622   153.78660523 259.18320897 202.22198093
 177.62015389 254.60843405 106.28554077  98.62946289 158.09650587
 114.9885088  218.09743995 136.7990638  210.97024658 123.06895
  78.48116316 118.72227208 165.97127875 117.25675494  93.99741376
 102.17605044 272.63201497 210.0372386  217.87035562 217.74094992
 191.77483615  77.49037697 200.62801612 104.23817549 115.41427184
 244.79188611 

In [14]:
print(y_test)

[ 79.  55. 109. 178.  39. 321. 158. 132.  52. 201.  95. 303.  96.  60.
  72.  60.  59.  73. 257.  84.  57. 222. 102.  72. 292. 134. 220. 198.
 275. 156. 232. 142. 277. 263. 191.  68.  37. 167.  90. 311.  42.  53.
 186. 202. 191. 151. 129.  85. 308. 248. 107. 243.  94.  47. 104.  63.
 248.  60. 248.  61.  48.  71. 122. 200.  49. 118. 243.  52. 296. 192.
 202.  59. 197.  94. 183. 306.  93. 332. 297.  85. 152.  72. 233. 172.
  72. 233. 281. 281. 121.]


**Here performance of model is seems very bad but it just tutorial how to make linear regression class of our own. so kindly ignore it...**

In [15]:
print(model.intercept_)
print(model.coef_)

#finding weights(coef_) and bius(intercept_) predicted by our model

152.88437994403927
[ -67.16308558 -244.0456999   509.15636468  305.70665091 -449.72246545
  217.82622156  -49.52742566   56.3929511   663.25035342   82.20307496]
